src/OrbitGl/MainThreadExecutor.h

```diff
@@ -79,16 +81,22 @@ class MainThreadExecutor : public std::enable_shared_from_this<MainThreadExecuto
     orbit_base::Promise<ReturnType> promise{};
     orbit_base::Future<ReturnType> resulting_future = promise.GetFuture();
 
-    auto continuation = [functor = std::forward<F>(functor), executor_weak_ptr = weak_from_this(),
+    waiting_continuations_.emplace_front(std::forward<F>(functor));
+    auto function_reference = waiting_continuations_.begin();
+
+    auto continuation = [this, function_reference, executor_weak_ptr = weak_from_this(),
                          promise = std::move(promise)](auto&&... argument) mutable {
       auto executor = executor_weak_ptr.lock();
       if (executor == nullptr) return;
 
       auto function_wrapper =
-          [functor = std::move(functor), promise = std::move(promise),
+          [this, function_reference, promise = std::move(promise),
            argument = std::make_tuple(std::forward<decltype(argument)>(argument)...)]() mutable {
             orbit_base::CallTaskAndSetResultInPromise<ReturnType> helper{&promise};
-            std::apply([&](auto... args) { helper.Call(functor, args...); }, std::move(argument));
+            auto functor = orbit_base::any_movable_cast<std::decay_t<F>>(&*function_reference);
+            CHECK(functor != nullptr);
+            std::apply([&](auto... args) { helper.Call(*functor, args...); }, std::move(argument));
+            waiting_continuations_.erase(function_reference);
           };
       executor->Schedule(CreateAction(std::move(function_wrapper)));
     };
```

Comment:
what happens if executor is destroyed after lock is released but before function_wrapper started execution?

Reply:
In that case `function_wrapper` will never be executed. That's guaranteed by `MainThreadExecutor::Schedule` since the lifetime of all actions are tied to the executor.

In our particular implementation we rely on the behaviour of `QMetaObject::invokeMethod`. The line `QMetaObject::invokeMethod(main_thread_executor_, function_wrapper, Qt::QueuedConnection);` will only invoke `function_wrapper` as long as `main_thread_executor_` is still alive.

So we can assume that either function_wrapper is called while MainThreadExecutor is still alive or MainThreadExecutor is destroyed and `function_wrapper` won't be called.

Reply:
ok, got it.

Reply:
I will add a comment explaining this.